e2e/test/scenarios/sharing/public-sharing-embed-button-behavior.cy.spec.js

```diff
@@ -254,14 +259,166 @@ describe("embed modal display", () => {
+          expectGoodSnowplowEvent({
+            event: "public_link_removed",
+            artifact: resource,
+            source: "public-share",
+          });
+        });
+      });
+
+      describe("when interacting with public embedding", () => {
+        it("should send `public_embed_code_copied` event when copying the public embed iframe", () => {
+          cy.get("@resourceId").then(id => {
+            visitResource(resource, id);
+          });
+
+          openEmbedModalFromMenu();
+          cy.findByTestId("sharing-pane-public-embed-button").within(() => {
+            cy.findByText("Get an embed link").click();
+          });
+          cy.findByTestId("copy-button").realClick();
+          expectGoodSnowplowEvent({
+            event: "public_embed_code_copied",
+            artifact: resource,
+            source: "public-embed",
+          });
+        });
+      });
+
+      describe("when interacting with static embedding", () => {
+        it("should send `static_embed_code_copied` when copying the static embed code", () => {
+          cy.get("@resourceId").then(id => {
+            visitResource(resource, id);
+          });
+          openStaticEmbeddingModal();
+          cy.findByTestId("embed-backend-code-sample").within(() => {
+            cy.findByTestId("copy-button").realClick();
+          });
+          expectGoodSnowplowEvent({
+            event: "static_embed_code_copied",
+            artifact: resource,
+            language: "Node.js",
+            location: "code_overview",
+            // appearance: {
+            //   "border": true,
+            //   "title": true,
+            //   "font": "instance",
+            //   "theme": "light"
+            // }
+          });
+        });
+
+        it("should send `static_embed_discarded` when discarding changes in the static embed modal", () => {
+          cy.get("@resourceId").then(id => {
+            enableEmbeddingForResource({ resource, id, isDirty: true });
+            visitResource(resource, id);
+          });
+
+          openStaticEmbeddingModal();
+
+          cy.findByTestId("embed-modal-content-status-bar").within(() => {
+            cy.findByText("Discard changes").click();
+          });
+
+          expectGoodSnowplowEvent({
+            event: "static_embed_discarded",
+            artifact: resource,
+          });
+        });
+
+        it("should send `static_embed_published` when publishing changes in the static embed modal", () => {
+          cy.get("@resourceId").then(id => {
+            visitResource(resource, id);
+          });
+          openStaticEmbeddingModal();
+
+          cy.findByTestId("embed-modal-content-status-bar").within(() => {
+            cy.findByText("Publish changes").click();
+          });
+
+          expectGoodSnowplowEvent({
+            event: "static_embed_published",
+            artifact: resource,
+            new_embed: true,
+            // params: {
```

Comment:
I'm having issues getting the `expectGoodSnowplowEvent` function to match with the nested properties. Anyone have any ideas?

Reply:
`expectGoodSnowplowEvent` uses `underscore.isMatch` https://github.com/metabase/metabase/blob/eb44d76cc3643db309add31baa750aee6afded54/e2e/support/helpers/e2e-snowplow-helpers.js#L32
That in turns uses `==` https://underscorejs.org/docs/modules/isMatch.html to compare the values, which doesn't work for objects

Reply:
I'll look into it 👀

Reply:
It seems that underscore doesn't have any way of doing a deep/recursive isMatch, the only "deep" function they have is `isEqual` but that fails is the first parameter has extra properties and that's not what we're doing at the moment.
Not sure if we should roll our own, it should be something like (this doesn't work for arrays but should work for the use cases we need at the moment)
```ts
const isDeepMatch = (obj, attrs) => {
  for (const [key, value] of Object.entries(attrs)) {
    if (!(key in obj)) {
      return false;
    }
    if (typeof value == "object") {
      if (!isDeepMatch(obj[key], value)) {
        return false;
      }
    } else {
      if (obj[key] !== value) {
        return false;
      }
    }
  }
  return true;
};
```